components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx

```diff
@@ -209,16 +211,15 @@ export class SidebarChannelMenu extends React.PureComponent<Props, State> {
                 return false;
             }
 
-            switch (channel.type) {
-            case Constants.OPEN_CHANNEL:
-            case Constants.PRIVATE_CHANNEL:
-                return category.type !== CategoryTypes.DIRECT_MESSAGES;
-            case Constants.DM_CHANNEL:
-            case Constants.GM_CHANNEL:
+            const selectedChannels = multiSelectedChannelIds.indexOf(channel.id) === -1 ? [channel] : displayedChannels.filter((c) => multiSelectedChannelIds.indexOf(c.id) !== -1);
+
+            if (selectedChannels.every((selectedChannel) => selectedChannel.type === Constants.DM_CHANNEL || selectedChannel.type === Constants.GM_CHANNEL)) {
                 return category.type !== CategoryTypes.CHANNELS;
-            default:
-                return true;
+            } else if (selectedChannels.every((selectedChannel) => selectedChannel.type !== Constants.DM_CHANNEL && selectedChannel.type !== Constants.GM_CHANNEL)) {
+                return category.type !== CategoryTypes.DIRECT_MESSAGES;
             }
```

Comment:
Sorry, i'm trying to grasp this, but seems I'm probably missing some context.

`multiSelectedChannelIds.indexOf(channel.id) === -1` is this trying to figure out if it is a multiselection or not?

I'm also not sure about this section of code, with many categories and channels selected this might slowdown. Is it going through every category and from there all the selected channels and then every channel selected in the category?

why do we check if it is a dm/gm selection only or a regular channel selection only after filtering by categories? wouldn't that give some errors? also if we can get that info before going throught the filter, we don't have to recalculate this for the whole selection twice. fixing this would probable not cause the problem i mentioned above, but as I said I might be missing something in this context.

Also it might be easier to read if we change the comparison from negatives to positives and maybe assigning that to a named constant, e.g.:

```js
selectedChannels.every((selectedChannel) => selectedChannel.type !== Constants.DM_CHANNEL && selectedChannel.type !== Constants.GM_CHANNEL)
```

vs

```js
const regularChannelsOnly = selectedChannels.every((selectedChannel) => selectedChannel.type === Constants.OPEN_CHANNEL || selectedChannel.type === Constants.PRIVATE_CHANNEL)
```
but this last part might be personal preference :)

Reply:
So the filtering here is done to only show categories in the menu that you can move the selected channels to.
The algorithm should be O(n^2) as it goes through all selectedChannels for each category to determine whether those selected channels can be moved to the category or not.

The only way I can think to speed this up would be to store this information in redux, but I'm not sure how much it would really speed it up. Might reduce the number of times it's recalculated if anything.

Thoughts?

Reply:
as discussed in community, probably moving the "isGM/DM" and "isRegularChannel" checks outside of the loop should improve everything

Reply:
@Willyfrog Ready for re-review!